node/libs/roles/src/validator/testonly.rs

```diff
@@ -62,6 +63,78 @@ pub fn make_block<R: Rng>(
     }
 }
 
+impl AggregateSignature {
+    /// Generate a new aggregate signature from a list of signatures.
+    pub fn aggregate<'a>(sigs: impl IntoIterator<Item = &'a Signature>) -> Self {
+        let mut agg = Self::default();
+        for sig in sigs {
+            agg.add(sig);
+        }
+        agg
+    }
+}
+
+impl PrepareQC {
+    /// Creates a new PrepareQC from a list of *signed* replica Prepare messages and the current validator set.
+    pub fn from(
+        signed_messages: &[Signed<ReplicaPrepare>],
+        validators: &ValidatorSet,
+    ) -> anyhow::Result<Self> {
+        // Get the view number from the messages, they must all be equal.
+        let view = signed_messages
+            .get(0)
+            .context("Empty signed messages vector")?
+            .msg
+            .view;
+
+        // Create the messages map.
+        let mut prepare_qc = PrepareQC::default();
+
+        for signed_message in signed_messages {
+            if signed_message.msg.view != view {
+                bail!("Signed messages aren't all for the same view.");
+            }
+
+            // Get index of the validator in the validator set.
+            let index = validators
+                .index(&signed_message.key)
+                .context("Message signer isn't in the validator set")?;
+
+            prepare_qc.add(signed_message, index, validators);
+        }
+
+        Ok(prepare_qc)
+    }
+}
+
+impl CommitQC {
+    /// Creates a new CommitQC from a list of *signed* replica Commit messages and the current validator set.
```

Comment:
... panics if signed_messages is empty.

Reply:
... expects signatures to be already verified.

Reply:
Fixed.